File: inverter_hil/python/inverterhilgui/format_inverter_panel.py
import math
from .gui_theme import guiTheme
from .format_measurement import formatMeasurement
from .format_torque_candidates import formatTorqueCandidates


def _numeric(item, field):
    value = item.get(field, math.nan) if isinstance(item, dict) else math.nan
    return float(value) if isinstance(value, (int, float)) and not isinstance(value, bool) else math.nan


def _flag(item, field, yes, no, fallback):
    value = item.get(field) if isinstance(item, dict) else None
    return yes if value is True or value == 1 else no if value is False or value == 0 else fallback


def formatInverterPanel(snapshot, index):
    theme = guiTheme(); nd = theme["text"]["noData"]
    panel = dict.fromkeys(("title", "corner", "state", "ready", "commandAge", "torqueCommand",
                           "torqueActual", "torqueCommandRaw", "torqueActualRaw", "speed", "idCurrent",
                           "iqCurrent", "motorTemperature", "switchTemperature", "derating", "activeFault"), nd)
    panel.update(corner=theme["text"]["cornerLabel"], cornerVerified=False, hasData=False)
    if not isinstance(index, int) or isinstance(index, bool) or index not in range(1, 5):
        return panel
    panel.update(title=f"INV{index}", corner=("FL", "FR", "RR", "RL")[index - 1], cornerVerified=True)
    items = snapshot.get("inverter") if isinstance(snapshot, dict) else None
    if not isinstance(items, (list, tuple)) or len(items) < index or not isinstance(items[index - 1], dict):
        return panel
    item = items[index - 1]
    state = item.get("state")
    panel["state"] = state if isinstance(state, str) and state.strip() else nd
    panel["ready"] = _flag(item, "ready", "READY", "NOT READY", nd)
    panel["derating"] = _flag(item, "derating", "ACTIVE", "NONE", nd)
    fault = item.get("activeFault")
    panel["activeFault"] = fault if isinstance(fault, str) and fault.strip() else "NONE"
    age, speed = _numeric(item, "commandAgeS"), _numeric(item, "speedRpm")
    panel["commandAge"] = f"{age:.3f} s" if math.isfinite(age) else nd
    panel["speed"] = f"{speed:.0f} rpm" if math.isfinite(speed) else nd
    for key, a, b in (("idCurrent", "idSetpointA", "idActualA"), ("iqCurrent", "iqSetpointA", "iqActualA")):
        av, bv = _numeric(item, a), _numeric(item, b)
        if math.isfinite(av) or math.isfinite(bv):
            panel[key] = f'{av:.1f}' if math.isfinite(av) else nd
            panel[key] += f' / {bv:.1f}' if math.isfinite(bv) else f' / {nd}'
            panel[key] += " A"
    command = formatTorqueCandidates(_numeric(item, "torqueCommandRaw"))
    actual = formatTorqueCandidates(_numeric(item, "torqueActualRaw"))
    panel.update(torqueCommand=command["summary"], torqueActual=actual["summary"],
                 torqueCommandRaw=command["raw"], torqueActualRaw=actual["raw"])
    motor = formatMeasurement(_numeric(item, "motorTemperatureC"), _numeric(item, "motorTemperatureRaw"), "C", True)
    switch = formatMeasurement(_numeric(item, "switchTemperatureC"), _numeric(item, "switchTemperatureRaw"), "C", False)
    panel.update(motorTemperature=motor["combined"], switchTemperature=switch["combined"])
    panel["hasData"] = command["hasData"] or actual["hasData"] or motor["hasData"] or switch["hasData"] or panel["state"] != nd
    return panel

File: VC_HIL/inverter/+inverterhilgui/formatInverterPanel.m
function panel = formatInverterPanel(snapshot, index)
%FORMATINVERTERPANEL Display fields for one inverter from a telemetry snapshot.
%
%   PANEL = FORMATINVERTERPANEL(SNAPSHOT, INDEX) reads only
%   SNAPSHOT.INVERTER(INDEX), so status from one inverter can never leak into
%   another panel.
%
%   CORNER MAPPING - operator-confirmed 2026-08-01:
%     Ephorus index 1 = FL, 2 = FR, 3 = RR, 4 = RL
%   Plan 4.1.3 held the corner label at UNVERIFIED until the FL/FR/RL/RR to
%   Ephorus-index mapping was confirmed; this is that confirmation, recorded
%   here as its single source. The provenance is an operator assertion, NOT an
%   automated check - nothing in this repository can verify which physical
%   wheel an index drives, so if that assertion is wrong every corner label is
%   wrong in a way no test will catch. Plan 4.1.3 also requires the canonical
%   inverter number to stay visible beside the corner, which PANEL.TITLE does.
%   CORNERVERIFIED reports whether CORNER is a real mapping or the UNVERIFIED
%   placeholder, so the caller can colour it accordingly.
%
%   Torque is rendered by INVERTERHILGUI.FORMATTORQUECANDIDATES, which always
%   shows the raw count and both candidate scales.

theme = inverterhilgui.guiTheme();
noData = theme.text.noData;

% Index order is Ephorus 1-4; see the corner-mapping note in the help above.
cornerNames = {'FL', 'FR', 'RR', 'RL'};

panel = struct( ...
    'title', noData, ...
    'corner', theme.text.cornerLabel, ...
    'cornerVerified', false, ...
    'state', noData, ...
    'ready', noData, ...
    'commandAge', noData, ...
    'torqueCommand', noData, ...
    'torqueActual', noData, ...
    'torqueCommandRaw', noData, ...
    'torqueActualRaw', noData, ...
    'speed', noData, ...
    'idCurrent', noData, ...
    'iqCurrent', noData, ...
    'motorTemperature', noData, ...
    'switchTemperature', noData, ...
    'derating', noData, ...
    'activeFault', noData, ...
    'hasData', false);

if ~isnumeric(index) || ~isscalar(index) || ~isreal(index) || ...
        ~isfinite(index) || index < 1 || index > 4 || index ~= floor(index)
    return;
end
panel.title = sprintf('INV%d', index);
% Only reached once INDEX is a validated 1-4, so the lookup cannot go out of
% range and an invalid index keeps the UNVERIFIED placeholder set above.
panel.corner = cornerNames{index};
panel.cornerVerified = true;
if ~isstruct(snapshot) || ~isscalar(snapshot) || ...
        ~isfield(snapshot, 'inverter') || numel(snapshot.inverter) < index
    return;
end
item = snapshot.inverter(index);

panel.state = textOr(item, 'state', noData);
panel.ready = flagText(item, 'ready', 'READY', 'NOT READY', noData);
panel.derating = flagText(item, 'derating', 'ACTIVE', 'NONE', noData);
panel.activeFault = textOr(item, 'activeFault', 'NONE');
if isempty(strtrim(panel.activeFault))
    panel.activeFault = 'NONE';
end
panel.commandAge = scalarText(item, 'commandAgeS', '%.3f s', noData);
panel.speed = scalarText(item, 'speedRpm', '%.0f rpm', noData);
panel.idCurrent = pairText(item, 'idSetpointA', 'idActualA', 'A', noData);
panel.iqCurrent = pairText(item, 'iqSetpointA', 'iqActualA', 'A', noData);

commandTorque = inverterhilgui.formatTorqueCandidates( ...
    numericOr(item, 'torqueCommandRaw'));
actualTorque = inverterhilgui.formatTorqueCandidates( ...
    numericOr(item, 'torqueActualRaw'));
panel.torqueCommand = commandTorque.summary;
panel.torqueActual = actualTorque.summary;
panel.torqueCommandRaw = commandTorque.raw;
panel.torqueActualRaw = actualTorque.raw;

motor = inverterhilgui.formatMeasurement( ...
    numericOr(item, 'motorTemperatureC'), ...
    numericOr(item, 'motorTemperatureRaw'), 'C', true);
switchTemperature = inverterhilgui.formatMeasurement( ...
    numericOr(item, 'switchTemperatureC'), ...
    numericOr(item, 'switchTemperatureRaw'), 'C', false);
panel.motorTemperature = motor.combined;
panel.switchTemperature = switchTemperature.combined;

panel.hasData = commandTorque.hasData || actualTorque.hasData || ...
    motor.hasData || switchTemperature.hasData || ...
    ~strcmp(panel.state, noData);
end

function text = textOr(item, field, fallback)
text = fallback;
if ~isfield(item, field)
    return;
end
value = item.(field);
if isstring(value) && isscalar(value)
    value = char(value);
end
if ischar(value) && ~isempty(strtrim(value))
    text = value;
end
end

function value = numericOr(item, field)
value = NaN;
if ~isfield(item, field)
    return;
end
candidate = item.(field);
if isnumeric(candidate) && isscalar(candidate) && isreal(candidate)
    value = double(candidate);
end
end

function text = flagText(item, field, trueText, falseText, noData)
text = noData;
if ~isfield(item, field)
    return;
end
value = item.(field);
if ~(islogical(value) || isnumeric(value)) || ~isscalar(value) || ...
        ~isreal(value) || ~isfinite(double(value)) || ...
        ~(double(value) == 0 || double(value) == 1)
    return;
end
if logical(value)
    text = trueText;
else
    text = falseText;
end
end

function text = scalarText(item, field, format, noData)
text = noData;
value = numericOr(item, field);
if isfinite(value)
    text = sprintf(format, value);
end
end

function text = pairText(item, setField, actualField, units, noData)
setValue = numericOr(item, setField);
actualValue = numericOr(item, actualField);
if ~isfinite(setValue) && ~isfinite(actualValue)
    text = noData;
    return;
end
text = sprintf('%s / %s %s', numberOrDash(setValue, noData), ...
    numberOrDash(actualValue, noData), units);
end

function text = numberOrDash(value, noData)
if isfinite(value)
    text = sprintf('%.1f', value);
else
    text = noData;
end
end
